hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetPlan.java

```diff
@@ -25,42 +26,165 @@
+    }
+
+    static AlterJobPlan toAlterJob(
+            PlanCacheKey id,
+            String jobName,
+            AlterJobOperation operation,
+            JetPlanExecutor planExecutor
+    ) {
+        return new AlterJobPlan(id, jobName, operation, planExecutor);
+    }
+
+    static CreateSnapshotPlan toCreateSnapshot(
+            PlanCacheKey id,
+            String snapshotName,
+            String jobName,
+            JetPlanExecutor planExecutor
+    ) {
+        return new CreateSnapshotPlan(id, snapshotName, jobName, planExecutor);
+    }
+
+    static DropJobPlan toDropJob(
+            PlanCacheKey id,
+            String jobName,
+            boolean ifExists,
+            String withSnapshotName,
+            JetPlanExecutor planExecutor
+    ) {
+        return new DropJobPlan(id, jobName, ifExists, withSnapshotName, planExecutor);
+    }
+
+    static DropSnapshotPlan toDropSnapshot(
+            PlanCacheKey id,
+            String snapshotName,
+            boolean ifExists,
+            JetPlanExecutor planExecutor
+    ) {
+        return new DropSnapshotPlan(id, snapshotName, ifExists, planExecutor);
+    }
+
+    static ShowStatementPlan toShowStatement(
+            PlanCacheKey id,
+            ShowStatementTarget showTarget,
+            JetPlanExecutor planExecutor
+    ) {
+        return new ShowStatementPlan(id, showTarget, planExecutor);
+    }
+
+    @SuppressWarnings("checkstyle:ParameterNumber")
+    static SelectOrSinkPlan toSelectOrSink(
+            PlanCacheKey id,
+            Set<PlanObjectKey> objectIds,
+            Map<UUID, PartitionIdSet> partitions,
+            DAG dag,
+            boolean isStreaming,
+            boolean isInsert,
+            SqlRowMetadata rowMetadata,
+            JetPlanExecutor planExecutor,
+            List<Permission> permissions
+    ) {
+        return objectIds.contains(PlanObjectKey.NON_CACHEABLE_OBJECT_ID)
+                ? new NonCacheableSelectOrSinkPlan(dag, isStreaming, isInsert, rowMetadata, planExecutor, permissions)
+                : new CacheableSelectOrSinkPlan(id, objectIds, partitions, dag, isStreaming, isInsert, rowMetadata,
+                planExecutor, permissions);
+    }
+
+    abstract class CacheableJetPlan implements JetPlan, CacheablePlan {
+
+        private final PlanCacheKey id;
+
+        private volatile long lastUsed;
+
+        protected CacheableJetPlan(PlanCacheKey id) {
+            this.id = id;
+        }
+
+        @Override
+        public final PlanCacheKey getPlanKey() {
+            assert id != null;
+
+            return id;
+        }
+
+        @Override
+        public final long getPlanLastUsed() {
+            return lastUsed;
+        }
+
+        @Override
+        public final void onPlanUsed() {
+            lastUsed = System.currentTimeMillis();
+        }
+    }
+
+    class CreateMappingPlan extends CacheableJetPlan {
```

Comment:
Maybe this and other DDL plans might be non-cacheable. It's very unusual if they're submitted repeatedly and then they just push out other plans out from the cache.